src/vs/vscode.d.ts

```diff
@@ -4587,6 +4623,110 @@ declare module 'vscode' {
 		 * Dispose and free associated resources.
 		 */
 		dispose(): void;
+
+		/**
+		 * Fires when the terminal's pty slave pseudo-device is written to. In other words, this
+		 * provides access to the raw data stream from the process running within the terminal,
+		 * including VT sequences.
+		 */
+		readonly onDidWriteData: Event<string>;
+	}
+
+	/**
+	 * Represents a terminal without a process where all interaction and output in the terminal is
+	 * controlled by an extension. This is similar to an output window but has the same VT sequence
+	 * compatibility as the regular terminal.
+	 *
+	 * Note that an instance of [Terminal](#Terminal) will be created when a TerminalRenderer is
+	 * created with all its APIs available for use by extensions. When using the Terminal object
+	 * of a TerminalRenderer it acts just like normal only the extension that created the
+	 * TerminalRenderer essentially acts as a process. For example when an
+	 * [Terminal.onDidWriteData](#Terminal.onDidWriteData) listener is registered, that will fire
+	 * when [TerminalRenderer.write](#TerminalRenderer.write) is called. Similarly when
+	 * [Terminal.sendText](#Terminal.sendText) is triggered that will fire the
+	 * [TerminalRenderer.onDidAcceptInput](#TerminalRenderer.onDidAcceptInput) event.
+	 *
+	 * **Example:** Create a terminal renderer, show it and write hello world in red
+	 * ```typescript
+	 * const renderer = window.createTerminalRenderer('foo');
+	 * renderer.terminal.then(t => t.show());
+	 * renderer.write('\x1b[31mHello world\x1b[0m');
+	 * ```
+	 */
+	export interface TerminalRenderer {
+		/**
+		 * The name of the terminal, this will appear in the terminal selector.
+		 */
+		name: string;
+
+		/**
+		 * The dimensions of the terminal, the rows and columns of the terminal can only be set to
+		 * a value smaller than the maximum value, if this is undefined the terminal will auto fit
+		 * to the maximum value [maximumDimensions](TerminalRenderer.maximumDimensions).
+		 *
+		 * **Example:** Override the dimensions of a TerminalRenderer to 20 columns and 10 rows
+		 * ```typescript
+		 * terminalRenderer.dimensions = {
+		 *   cols: 20,
+		 *   rows: 10
+		 * };
+		 * ```
+		 */
+		dimensions: TerminalDimensions | undefined;
+
+		/**
+		 * The maximum dimensions of the terminal, this will be undefined immediately after a
+		 * terminal renderer is created and also until the terminal becomes visible in the UI.
+		 * Listen to [onDidChangeMaximumDimensions](TerminalRenderer.onDidChangeMaximumDimensions)
+		 * to get notified when this value changes.
+		 */
+		readonly maximumDimensions: TerminalDimensions | undefined;
+
+		/**
+		 * The corresponding [Terminal](#Terminal) for this TerminalRenderer.
+		 */
+		readonly terminal: Terminal;
```

Comment:
Can you explain there terminal and terminal-renderer relation? Why does the renderer know the terminal and why doesn't a terminal know the renderer?

Reply:
A `TerminalRenderer` will always have a `Terminal`, a `Terminal` will not necessarily have a `TerminalRenderer` though. Additionally, only extensions that created a `TerminalRenderer` via `createTerminalRenderer` will ever have access to them, this is important because `TerminalRenderer`s allow the extension to listen to terminal input which could be passwords, tokens, etc.

The reason for the `TerminalRenderer.terminal` property is to give convenient access to members on `Terminal` so they don't need to be duplicated, such as `Terminal.show`, `Terminal.hide` and `Terminal.dispose`. We could also duplicate those members in `TerminalRenderer` if that's preferable:

```ts
export interface TerminalRenderer {
	/**
	 * Show the terminal panel and reveal this terminal in the UI.
	 *
	 * @param preserveFocus When `true` the terminal will not take focus.
	 */
	show(preserveFocus?: boolean): void;

	/**
	 * Hide the terminal panel if this terminal is currently showing.
	 */
	hide(): void;

	/**
	 * Dispose and free associated resources.
	 */
	dispose(): void;
}
```

The only thing I think we would lose if we went this route is the ability to reliably determine which `Terminal`s have `TerminalRenderer`s, say if an extension was doing something with all `Terminal`s that it did not create via `createTerminalRenderer`. Probably not a big deal losing that.

Reply:
> The reason for the `TerminalRenderer.terminal` property is to give convenient access to members on `Terminal` so they don't need to be duplicated, such as `Terminal.show`, `Terminal.hide` and `Terminal.dispose`.

That means the terminal domain doesn't mandate that, right? A terminal and a terminal render just happen to have some overlap in the interface and they happen to be showing in the same UX space, right? Iff so then I would strongly prefer to decouple them so that confusion is avoided. Else question about `terminalRenderer.terminal.processId` arise and it might make it harder to evolve the terminal without evolving terminal renderer. 

> have access to them, this is important because `TerminalRenderer`s allow the extension to listen to terminal input which could be passwords, tokens, etc.

Fake security? If I am an extension and interested in your passwords I won't wait for you typing them in a terminal.

Reply:
> Iff so then I would strongly prefer to decouple them so that confusion is avoided.

Sounds good to me 👍